src/v/kafka/server/tests/quota_manager_test.cc

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "kafka/server/quota_manager.h"
+#include "random/generators.h"
+#include "resource_mgmt/rate.h"
+
+#include <seastar/core/lowres_clock.hh>
+
+#include <boost/test/tools/old/interface.hpp>
+#include <boost/test/unit_test.hpp>
+
+#include <chrono>
+
+using namespace std::chrono_literals;
+
+void set_quota_manager_test_defaults(
+  int16_t num_windows,
+  std::chrono::milliseconds window_size,
+  uint32_t quota_rate,
+  std::chrono::milliseconds max_delay) {
+    /// Configure quota manager with known defaults otherwise if assumptions are
+    /// made on tests defaults values obtained from config settings, test would
+    /// fail if these defaults are later changed
+    auto& cfg = config::shard_local_cfg();
+    cfg.get("default_num_windows").set_value(num_windows);
+    cfg.get("default_window_sec").set_value(window_size);
+    cfg.get("target_quota_time_rate")
+      .set_value(std::optional<uint32_t>(quota_rate));
+    cfg.get("max_kafka_throttle_delay_ms").set_value(max_delay);
+}
+
+/// This test emmulates a single client making frequent requests to
+/// redpanda. Each request is given a random time interval (between 1-10ms),
+/// logged in the quota manager and tracked by a local instance of
+/// rate_tracker
+BOOST_AUTO_TEST_CASE(quota_manager_test) {
+    const auto num_windows = 10;
+    const auto window_size = 1000ms;
+    const auto quota_rate = 80;
```

Comment:
Does a quota rate of 100 effectively mean there's no time-based throttling?

Reply:
As per the [kip](https://cwiki.apache.org/confluence/display/KAFKA/KIP-124+-+Request+rate+quotas) the quota is the percentage of allowable time within a window clients are allowed to occupy. So a quota rate of 100 means the full window here, 1000ms. To disable this feature set the initial configuration option to nullopt in `configuration.cc`